File: solution-python/036_valid_sudoku.py
"""
	036. Valid Sudoku
	@name: li jin
	@date: Apr 24, 2016
	@link: https://leetcode.com/problems/valid-sudoku/
	@time: 88 ms
"""

class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        tt = [[[False for i in range(9)] for j in range(9)] for k in range(3)]
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    idx = ord(board[i][j]) - ord('1')
                    k = i / 3 * 3 + j / 3
                    if tt[0][i][idx] or tt[1][j][idx] or tt[2][k][idx]:
                        return False
                    tt[0][i][idx] = True
                    tt[1][j][idx] = True
                    tt[2][k][idx] = True
        return True

File: solution-cpp/036_valid_sudoku.cpp
/*
	036. Valid Sudoku
	@name: li jin
	@date: Apr 24, 2016
	@link: https://leetcode.com/problems/valid-sudoku/
	@time: 12 ms
*/

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        bool nine[3][9][9] = {false};
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    int n = board[i][j] - '1';
                    int k = (i/3) * 3 + j / 3;
                    if (nine[0][i][n] || nine[1][j][n] || nine[2][k][n])
                        return false;
                    nine[0][i][n] = true;
                    nine[1][j][n] = true;
                    nine[2][k][n] = true;
                }
            }
        }
        return true;
    }
};
